File: run_source.py
"""Compile and execute minimal MIN0 CORE FORTH source."""

from __future__ import annotations

import argparse
import json
from pathlib import Path

from min0_core_forth_compiler import compile_source
from min0_core_forth_vm import Min0CoreForthVM


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("source", type=Path)
    parser.add_argument("--write-image", type=Path)
    args = parser.parse_args()
    image = compile_source(args.source.read_text(encoding="utf-8"))
    if args.write_image:
        args.write_image.write_bytes(image)
    vm = Min0CoreForthVM()
    vm.load(image)
    stack = vm.run()
    print(
        json.dumps(
            {"stack": stack, "steps": vm.steps, "image_bytes": len(image)},
            separators=(",", ":"),
        )
    )


if __name__ == "__main__":
    main()

File: run_source.rb
# frozen_string_literal: true

require "json"
require_relative "min0_core_forth_compiler"

abort "usage: ruby run_source.rb SOURCE [IMAGE]" unless (1..2).cover?(ARGV.length)

image = Min0CoreForth::Compiler.compile(File.read(ARGV.fetch(0), encoding: "UTF-8"))
File.binwrite(ARGV.fetch(1), image) if ARGV[1]
vm = Min0CoreForth::VM.new
vm.load(image)
stack = vm.run
puts JSON.generate({ stack: stack, steps: vm.steps, image_bytes: image.bytesize })
